components/automate-gateway/handler/iam/v2beta/policy/policy.go

```diff
@@ -424,6 +430,62 @@ func (p *Server) ListProjects(
 	return &pb_resp.ListProjectsResp{Projects: projects}, nil
 }
 
+// IntrospectAllProjects returns the list of all projects for display in the
+// global projects filter, so it may include unassigned.
+func (p *Server) IntrospectAllProjects(
+	ctx context.Context, req *pb_req.ListProjectsReq) (*pb_resp.ListProjectsResp, error) {
+
+	projectIDs, err := p.getAllowedProjects(ctx)
+	if err != nil {
+		return nil, err
+	}
+	// To correlate names with project IDs, use the standard ListProjects
+	// with the special flag to include "(unassigned)" too.
+	resp, err := p.projects.ListProjects(ctx, &authz.ListProjectsReq{GlobalFilterView: true})
+	if err != nil {
+		return nil, err
+	}
+	return mapProjectNamesToIds(projectIDs, resp.Projects)
+}
+
+func mapProjectNamesToIds(
+	projectIDs []string, refProjects []*authz.Project) (*pb_resp.ListProjectsResp, error) {
+	// refProjects is *all* projects, not just all *allowed* projects.
+	// This uses the correctly filtered id list to generate the desired return list.
+	projMap := toMap(refProjects)
+	projects := make([]*pb_common.Project, len(projectIDs))
+	for i, id := range projectIDs {
+		apiProj, err := domainProjectToAPIProject(projMap[id])
```

Comment:
This could fail, I believe. `projMap[id]` will be `nil` if no project with that id is in `refProjects`. Now, I know that this should be highly unlikely, but when facing concurrent user/API requests, this could perhaps still happen.
```suggestion
		domainProj, ok := projMap[id]
		if !ok {
			// deal with it
		}
		apiProj, err := domainProjectToAPIProject(domainProj)
```

Reply:
Thanks for that; good point.